FirebaseAppDistribution/Sources/FIRAppDistribution.m

```diff
@@ -0,0 +1,291 @@
+
+// The App Distribution Tester API endpoint used to retrieve releases
+NSString *const ReleasesEndpointURL =
+    @"https://firebaseapptesters.googleapis.com/v1alpha/devices/-/testerApps/%@/releases";
+
+@synthesize isTesterSignedIn = _isTesterSignedIn;
+
+#pragma mark - Singleton Support
+
+- (instancetype)initWithApp:(FIRApp *)app appInfo:(NSDictionary *)appInfo {
+  self = [super init];
+
+  if (self) {
+    self.safariHostingViewController = [[UIViewController alloc] init];
+
+    // Save any properties here
+    NSLog(@"APP DISTRIBUTION STARTED UP!");
+
+    [GULAppDelegateSwizzler proxyOriginalDelegate];
+
+    FIRAppDistributionAppDelegatorInterceptor *interceptor =
+        [FIRAppDistributionAppDelegatorInterceptor sharedInstance];
+    [GULAppDelegateSwizzler registerAppDelegateInterceptor:interceptor];
+  }
+
+  // TODO: Lookup keychain to load auth state on init
+  _isTesterSignedIn = self.authState ? YES : NO;
+  return self;
+}
+
++ (void)load {
+  [FIRApp registerInternalLibrary:(Class<FIRLibrary>)self
+                         withName:@"firebase-appdistribution"
+                      withVersion:@"0.0.0"];  // TODO: Get version from podspec
+}
+
++ (NSArray<FIRComponent *> *)componentsToRegister {
+  FIRComponentCreationBlock creationBlock =
+      ^id _Nullable(FIRComponentContainer *container, BOOL *isCacheable) {
+    if (!container.app.isDefaultApp) {
+      NSLog(@"App Distribution must be used with the default Firebase app.");
+      return nil;
+    }
+
+    *isCacheable = YES;
+
+    return [[FIRAppDistribution alloc] initWithApp:container.app
+                                           appInfo:NSBundle.mainBundle.infoDictionary];
+  };
+
+  FIRComponent *component =
+      [FIRComponent componentWithProtocol:@protocol(FIRAppDistributionInstanceProvider)
+                      instantiationTiming:FIRInstantiationTimingEagerInDefaultApp
+                             dependencies:@[]
+                            creationBlock:creationBlock];
+  return @[ component ];
+}
+
++ (instancetype)appDistribution {
+  // The container will return the same instance since isCacheable is set
+
+  FIRApp *defaultApp = [FIRApp defaultApp];  // Missing configure will be logged here.
+
+  // Get the instance from the `FIRApp`'s container. This will create a new instance the
+  // first time it is called, and since `isCacheable` is set in the component creation
+  // block, it will return the existing instance on subsequent calls.
+  id<FIRAppDistributionInstanceProvider> instance =
+      FIR_COMPONENT(FIRAppDistributionInstanceProvider, defaultApp.container);
+
+  // In the component creation block, we return an instance of `FIRAppDistribution`. Cast it and
+  // return it.
+  return (FIRAppDistribution *)instance;
+}
+
+- (void)signInTesterWithCompletion:(FIRAppDistributionSignInTesterCompletion)completion {
+  NSURL *issuer = [NSURL URLWithString:@"https://accounts.google.com"];
```

Comment:
Set all the urls as constants at the top of the file?

Reply:
Done